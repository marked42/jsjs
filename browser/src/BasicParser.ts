export class BasicParser {
  private index = 0
  constructor(private input: string) {}

  char() {
    return this.input[this.index]
  }

  next(len?: number) {
    len = len || 1
    this.index += len
  }

  eat(char?: string) {
    if (typeof char === 'string') {
      this.assert(char)
    }

    const text = char ? char : this.char()
    this.next(text ? text.length : 1)
    return text
  }

  assert(char: string) {
    if (!this.input.substring(this.index).startsWith(char)) {
      throw new Error(`unexpected char ${char}`)
    }
  }

  consumeWhitespace() {
    const whitespace = /[\t|\r|\n| ]/
    const startIndex = this.index
    while (whitespace.test(this.char())) {
      this.eat()
    }

    return this.input.substring(startIndex, this.index)
  }

  startsWith(input: string) {
    return this.input.substring(this.index).indexOf(input) === 0
  }

  end() {
    return this.index >= this.input.length
  }

  consumeCharsWhile(predicate: (char: string) => boolean) {
    let text = ''
    while (predicate(this.char()) && !this.end()) {
      text += this.eat()
    }

    return text
  }
}
