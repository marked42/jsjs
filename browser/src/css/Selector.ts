export class Selector {
  constructor(
    public classes: string[],
    public tagName?: string,
    public id?: string
  ) {}
}
