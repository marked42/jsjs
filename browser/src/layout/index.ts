import {
  DeclarationValue,
  DeclarationValueKeyword,
  DeclarationValueLength,
  Unit,
} from '../css'
import { StyleNode } from '../style'
import { Dimensions } from './Dimensions'

export enum LayoutBoxType {
  Block,
  Inline,
  Anonymous,
}

export class LayoutBox {
  public dimensions = Dimensions.default()
  public children = [] as LayoutBox[]
  constructor(public boxType: LayoutBoxType, public node?: StyleNode) {}

  getInlineContainer() {
    switch (this.boxType) {
      case LayoutBoxType.Inline:
      case LayoutBoxType.Anonymous:
        return this
      case LayoutBoxType.Block:
        const lastChild = this.children[this.children.length - 1]
        if (lastChild && lastChild.boxType === LayoutBoxType.Anonymous) {
          return lastChild
        } else {
          const anonymousBox = new LayoutBox(LayoutBoxType.Anonymous)

          this.children.push(anonymousBox)
          return anonymousBox
        }
    }
  }

  layout(containingBlock: Dimensions) {
    switch (this.boxType) {
      case LayoutBoxType.Block:
        this.layoutBlock(containingBlock)
      case LayoutBoxType.Inline:
      case LayoutBoxType.Anonymous:
        break
    }
  }

  layoutBlock(containingBlock: Dimensions) {
    this.calculateBlockWidth(containingBlock)
    this.calculateBlockPosition(containingBlock)
    this.layoutBlockChildren()
    this.calculateBlockHeight(containingBlock)
  }

  calculateBlockWidth(containingBlock: Dimensions) {
    const node = this.getStyleNode()

    const auto = DeclarationValueKeyword.of('auto')
    let width = node.value('width') || auto

    const zero = new DeclarationValueLength(0, Unit.Px)
    let marginLeft = node.lookup('margin-left', 'margin', zero)
    const borderLeft = node.lookup('border-left-width', 'border-width', zero)
    const paddingLeft = node.lookup('padding-left', 'padding', zero)
    let marginRight = node.lookup('margin-right', 'margin', zero)
    const borderRight = node.lookup('border-right-width', 'border-width', zero)
    const paddingRight = node.lookup('padding-right', 'padding', zero)

    const toPx = (value: DeclarationValue) => {
      if (value instanceof DeclarationValueLength) {
        return value.value
      }

      if (value instanceof DeclarationValueKeyword && value.value === 'auto') {
        return 0
      }

      throw new Error(`unexpected value ${value}`)
    }

    const totalWidth = [
      marginLeft,
      borderLeft,
      paddingLeft,
      width,
      paddingRight,
      borderRight,
      marginRight,
    ].reduce((acc, value) => {
      return acc + toPx(value)
    }, 0)

    const underflow = containingBlock.content.width - totalWidth

    // 把auto值转换为固定长度值
    // overflow
    if (underflow < 0) {
      if (marginLeft === auto) {
        marginLeft = new DeclarationValueLength(0, Unit.Px)
      }
      if (marginRight === auto) {
        marginRight = new DeclarationValueLength(0, Unit.Px)
      }
      if (width === auto) {
        width = new DeclarationValueLength(0, Unit.Px)
      } else {
        marginRight = new DeclarationValueLength(
          toPx(marginRight) + underflow,
          Unit.Px
        )
      }
    } else {
      // underflow 的8种情况
      if (width !== auto) {
        if (marginLeft === auto && marginRight === auto) {
          marginLeft = marginRight = new DeclarationValueLength(
            0.5 * underflow,
            Unit.Px
          )
        } else if (marginLeft === auto && marginRight !== auto) {
          marginLeft = new DeclarationValueLength(underflow, Unit.Px)
        } else if (marginLeft !== auto && marginRight === auto) {
          marginRight = new DeclarationValueLength(underflow, Unit.Px)
        } else if (marginLeft !== auto && marginRight !== auto) {
          marginRight = new DeclarationValueLength(
            toPx(marginRight) + underflow,
            Unit.Px
          )
        }
      } else {
        width = new DeclarationValueLength(underflow, Unit.Px)
        if (marginLeft === auto) {
          marginLeft = new DeclarationValueLength(0, Unit.Px)
        }
        if (marginRight === auto) {
          marginRight = new DeclarationValueLength(0, Unit.Px)
        }
      }
    }
  }

  getStyleNode() {
    if (this.boxType === LayoutBoxType.Anonymous || !this.node) {
      throw new Error('anonymous box has no style node')
    }

    return this.node
  }

  calculateBlockPosition(containingBlock: Dimensions) {}

  layoutBlockChildren() {
    this.children.forEach((child) => child.layoutBlock(this.dimensions))
  }

  calculateBlockHeight(containingBlock: Dimensions) {}
}

export function getLayoutBoxType(node: StyleNode) {
  const display = node.specifiedValues['display']

  if (display instanceof DeclarationValueKeyword) {
    if (display.value === 'block') {
      return LayoutBoxType.Block
    } else if (display.value === 'none') {
      return LayoutBoxType.Inline
    } else {
      return LayoutBoxType.Block
    }
  }
  return LayoutBoxType.Block
}

export function getDisplayValue(node: StyleNode) {
  const display = node.specifiedValues['display']

  if (display instanceof DeclarationValueKeyword) {
    if (display.value === 'block') {
      return 'block'
    } else if (display.value === 'none') {
      return 'none'
    } else {
      return 'inline'
    }
  }
  return 'inline'
}

export function buildLayoutTree(node: StyleNode) {
  const display = getDisplayValue(node)
  const root = new LayoutBox(LayoutBoxType.Block, node)
  switch (display) {
    case 'inline':
      root.boxType = LayoutBoxType.Inline
      break
    case 'block':
      root.boxType = LayoutBoxType.Block
      break
    case 'none':
      throw new Error('display none at root node not allowed')
  }

  node.children.forEach((child) => {
    const childDisplay = getDisplayValue(child)
    switch (childDisplay) {
      case 'block':
        root.children.push(buildLayoutTree(child))
        break
      case 'inline':
        root.getInlineContainer().children.push(buildLayoutTree(child))
      case 'none':
        break
    }
  })

  return root
}
