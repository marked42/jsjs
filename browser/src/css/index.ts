import { BasicParser } from '../BasicParser'

class Stylesheet {
  constructor(public rules: Rule[]) {}
}

class Rule {
  constructor(
    public selectors: Selector[],
    public declarations: Declaration[]
  ) {}
}

class Selector {
  constructor(
    public tagName: string,
    public id: string,
    public classes: string[]
  ) {}
}

class Declaration {
  constructor(public name: string, public value: string) {}
}

enum DeclarationValueType {
  Keyword,
  Length,
  Color,
}

class DeclarationValue {
  constructor(public type: DeclarationValueType) {}
}

class DeclarationValueKeyword extends DeclarationValue {
  constructor(public value: string) {
    super(DeclarationValueType.Keyword)
  }
}

class DeclarationValueLength extends DeclarationValue {
  constructor(public value: number, public unit: Unit) {
    super(DeclarationValueType.Length)
  }
}

enum Unit {
  Px,
}

interface Color {
  r: number
  g: number
  b: number
  a: number
}

class DeclarationValueColor extends DeclarationValue {
  constructor(public value: Color) {
    super(DeclarationValueType.Color)
  }
}

export class CSSParser extends BasicParser {
  parseCSS() {}
}

export function parseCSS(input: string) {
  const parser = new CSSParser(input)
  return parser.parseCSS()
}
