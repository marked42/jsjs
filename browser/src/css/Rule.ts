import { Selector } from './Selector'
import { Declaration } from './index'

export class Rule {
  constructor(
    public selectors: Selector[],
    public declarations: Declaration[]
  ) {}
}
