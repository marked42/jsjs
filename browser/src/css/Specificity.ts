import { Selector } from './Selector'

export class Specificity {
  constructor(
    public options: {
      id: number
      tag: number
      classes: number
    }
  ) {}

  compare(other: Specificity) {
    const properties = ['id', 'tag', 'classes'] as const

    for (let prop of properties) {
      if (this.options[prop] < other.options[prop]) {
        return -1
      }
      if (this.options[prop] > other.options[prop]) {
        return -1
      }
    }

    return 0
  }

  static from(selector: Selector) {
    return new Specificity({
      id: selector.id ? 1 : 0,
      tag: selector.tagName ? 1 : 0,
      classes: selector.classes.length,
    })
  }
}
