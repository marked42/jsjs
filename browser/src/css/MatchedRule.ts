import { Rule } from './Rule'
import { Specificity } from './Specificity'

export class MatchedRule {
  constructor(public specificity: Specificity, public rule: Rule) {}
}
