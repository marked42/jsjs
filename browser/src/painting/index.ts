import { LayoutBox, LayoutBoxType } from '../layout'
import { Rect } from '../layout/Rect'
import { Color, DeclarationValueColor } from '../css'

class DisplayCommand {}

class DisplayCommandSolidColor extends DisplayCommand {
  constructor(public color: Color, public rect: Rect) {
    super()
  }
}

export function buildDisplayList(layoutBox: LayoutBox): DisplayCommand[] {
  const displayList = [] as DisplayCommand[]
  renderLayoutBox(displayList, layoutBox)
  return displayList
}

export function renderLayoutBox(list: DisplayCommand[], layoutBox: LayoutBox) {
  renderBackground(list, layoutBox)
  renderBorders(list, layoutBox)

  layoutBox.children.forEach((child) => {
    renderLayoutBox(list, child)
  })
}

export function renderBackground(list: DisplayCommand[], layoutBox: LayoutBox) {
  const backgroundColor = getBoxColor(layoutBox, 'background')
  if (!backgroundColor) {
    return
  }

  list.push(
    new DisplayCommandSolidColor(
      backgroundColor,
      layoutBox.dimensions.borderBox()
    )
  )
}

export function renderBorders(list: DisplayCommand[], layoutBox: LayoutBox) {
  const borderColor = getBoxColor(layoutBox, 'border-color')
  if (!borderColor) {
    return
  }

  const dimensions = layoutBox.dimensions
  const borderBox = layoutBox.dimensions.borderBox()

  // left
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x,
        borderBox.y,
        dimensions.border.left,
        borderBox.height
      )
    )
  )

  // right
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x +
          dimensions.border.left +
          dimensions.padding.left +
          dimensions.content.width,
        borderBox.y,
        dimensions.border.right,
        borderBox.height
      )
    )
  )

  // top border
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(borderBox.x, borderBox.y, dimensions.border.top, borderBox.width)
    )
  )

  // bottom color
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x,
        borderBox.y + borderBox.height - dimensions.border.bottom,
        borderBox.width,
        dimensions.border.bottom
      )
    )
  )
}

export function getBoxColor(layoutBox: LayoutBox, name: string) {
  switch (layoutBox.boxType) {
    case LayoutBoxType.Block:
    case LayoutBoxType.Inline:
      const styleNode = layoutBox.getStyleNode()
      const value = styleNode.value(name)
      if (value) {
        if (value instanceof DeclarationValueColor) {
          return value.value
        } else {
          throw new Error(`style ${name} value is not color, get ${value}|`)
        }
      }
    case LayoutBoxType.Anonymous:
    default:
      break
  }
}

interface CanvasData {
  data: Uint8ClampedArray
  width: number
  height: number
}

export function drawLayoutBoxAsCanvasData(layoutBox: LayoutBox, bounds: Rect) {
  const canvasData = {
    data: new Uint8ClampedArray(4 * bounds.width * bounds.height),
    width: bounds.width,
    height: bounds.height,
  }

  const displayCommandList = buildDisplayList(layoutBox)

  for (const command of displayCommandList) {
    if (command instanceof DisplayCommandSolidColor) {
      applyCommandToCanvasData(command, canvasData)
    }
  }

  return canvasData
}

function applyCommandToCanvasData(
  command: DisplayCommandSolidColor,
  canvasData: CanvasData
) {
  for (let row = 0; row < canvasData.height; row++) {
    for (let col = 0; col < canvasData.width; col++) {
      canvasData.data[row * canvasData.width + col] = command.color.r
      canvasData.data[row * canvasData.width + col + 1] = command.color.g
      canvasData.data[row * canvasData.width + col + 2] = command.color.b
      canvasData.data[row * canvasData.width + col + 3] = command.color.a
    }
  }
}
