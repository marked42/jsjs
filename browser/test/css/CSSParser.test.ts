import { parseCSS } from '../../src/css'

describe('parse css', () => {
  it('should parse css', () => {
    const input = `
	div#test.first , test {
		font-size: 14px;
		font-color: #ffffffff;
		font-weight: bold;
	}

	first , test {
		font-size: 14px;
		font-color: #ffffffff;
		font-weight: bold;
	}
		`
    expect(parseCSS(input)).toMatchSnapshot()
  })
})
