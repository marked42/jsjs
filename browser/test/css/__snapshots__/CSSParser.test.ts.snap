// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse css should parse css 1`] = `
StyleSheet {
  "rules": Array [
    Rule {
      "declarations": Array [
        Declaration {
          "name": "font-size",
          "value": DeclarationValueLength {
            "type": 1,
            "unit": 0,
            "value": 14,
          },
        },
        Declaration {
          "name": "font-color",
          "value": DeclarationValueColor {
            "type": 2,
            "value": Object {
              "a": 255,
              "b": 255,
              "g": 255,
              "r": 255,
            },
          },
        },
        Declaration {
          "name": "font-weight",
          "value": DeclarationValueKeyword {
            "type": 0,
            "value": "bold",
          },
        },
      ],
      "selectors": Array [
        Selector {
          "classes": Array [
            "first",
          ],
          "id": "test",
          "tagName": "div",
        },
        Selector {
          "classes": Array [],
          "id": undefined,
          "tagName": "test",
        },
      ],
    },
    Rule {
      "declarations": Array [
        Declaration {
          "name": "font-size",
          "value": DeclarationValueLength {
            "type": 1,
            "unit": 0,
            "value": 14,
          },
        },
        Declaration {
          "name": "font-color",
          "value": DeclarationValueColor {
            "type": 2,
            "value": Object {
              "a": 255,
              "b": 255,
              "g": 255,
              "r": 255,
            },
          },
        },
        Declaration {
          "name": "font-weight",
          "value": DeclarationValueKeyword {
            "type": 0,
            "value": "bold",
          },
        },
      ],
      "selectors": Array [
        Selector {
          "classes": Array [],
          "id": undefined,
          "tagName": "first",
        },
        Selector {
          "classes": Array [],
          "id": undefined,
          "tagName": "test",
        },
      ],
    },
  ],
}
`;
